Create DS interface and DS factory classes

domire8 opened the pull request:
Sooo this is the first PR in a series of PRs to refactor the dynamical systems module. I'm adding here the interface class as well as the factory. I hope this is what you guys were looking for. The DS will from now on have the following interface methods:
- get/set_parameter
- get/set_base_frame
- is_compatible
- evaluate

Let me know if thats what you expected from this refactor. I have everything ready in the background (ring, circular, linear DS), just to mention that this design works :)

eeberhard commented on source/dynamical_systems/src/IDynamicalSystem.cpp:
```diff
@@ -0,0 +1,138 @@
+
+template<class S>
+void IDynamicalSystem<S>::set_parameter(const std::shared_ptr<ParameterInterface>&) {
+  throw exceptions::NotImplementedException("set_parameter() not implemented.");
```
This implementation should also be done here, so mark it with a TODO. It would need to `get_name` from the passed parameter, `assert_parameter_exists(name)`, and then update the value in the map through the shared ptr ParameterInterface.

eeberhard reviewed (commented):
Looks pretty good, though I have a few comments. Let me know if you want to discuss any of it in more detail.

eeberhard commented on source/dynamical_systems/src/IDynamicalSystem.cpp:
```diff
@@ -0,0 +1,138 @@
+}
+
+template std::list<std::shared_ptr<ParameterInterface>> IDynamicalSystem<CartesianState>::get_parameters() const;
+template std::list<std::shared_ptr<ParameterInterface>> IDynamicalSystem<JointState>::get_parameters() const;
```
Turns out I was wrong about this ([CI failure](https://github.com/epfl-lasa/control-libraries/runs/4225639832?check_suite_focus=true))! Certain template things still catch me out. So the declaration of the specialization does have to come afterwards apparently...

buschbapti commented on source/dynamical_systems/include/dynamical_systems/IDynamicalSystem.hpp:
```diff
@@ -0,0 +1,114 @@
+   * @brief Get a list of all the parameters of the dynamical system.
+   * @return The list of parameters
+   */
+  [[nodiscard]] std::list<std::shared_ptr<state_representation::ParameterInterface>> get_parameters() const;
```
In `modulo` parameters are actually stored as a `map` where the key is the name or the parameter. In fact this could return a `map` as well but the reason I opted for a list here is that you just need to go trough all the parameters returned by the class and store them in a `map` at the `Cell` level. The is a function `add_parameters` in `Cell` that takes this list, declare them in `ros` and store them in the `map`. Obviously this function could also take a map if that is preferable,

eeberhard replied:
Then we should just return the map here, that will be easier for everyone. It will already be a breaking change anyway if this is released as 5.0

buschbapti commented on source/dynamical_systems/include/dynamical_systems/DefaultDynamicalSystem.hpp:
```diff
@@ -0,0 +1,27 @@
+   * @param state The input state
+   * @return The output state
+   */
+  [[nodiscard]] S compute_dynamics(const S& state) const override;
```
I guess this makes sense to declare the `const` function as `nodiscard`. There is a lot of those we didn't do yet though. Should we make it the new pattern from now on?

eeberhard replied:
My opinion that I discussed with dominic is that we should use this marking when convenient, since it is otherwise a soft warning in C++17, but I don't mind if it is absent in some places (so it's not something that necessitates a refactor, but for new code we might as well)

buschbapti commented on source/dynamical_systems/include/dynamical_systems/DynamicalSystemFactory.hpp:
```diff
@@ -0,0 +1,29 @@
+  /**
+   * @brief Enumeration of the implemented dynamical systems
+   */
+  enum DYNAMICAL_SYSTEM {
```
I believe a `class enum` would be preferred here (see https://stackoverflow.com/questions/18335861/why-is-enum-class-preferred-over-plain-enum for the reasons)

buschbapti reviewed (commented):
This goes in the good direction indeed. I agree with @eeberhard comments.

buschbapti commented on source/dynamical_systems/include/dynamical_systems/IDynamicalSystem.hpp:
```diff
@@ -0,0 +1,114 @@
+   * @param state State at which to perform the evaluation
+   * @return The resulting state (velocity) of the dynamical system
+   */
+  [[nodiscard]] S evaluate(const S& state) const;
```
That is a good point. I am not against remove the `const` actually if you think that make more sense. Adding the `const chronotype& dt` could also be a nice solution. Both this and `compute_dynamics` should be treated in the same way I think.

eeberhard replied:
I think two methods, a const like this and non-const with the additional argument, will be a nice solution. Since `compute_dynamics` is the function to be overridden, you are correct that it needs to have the equivalent const and non-const variants. This can wait for a future PR though.

domire8 replied:
Okay so nothing to do here for me, at least for now?

domire8 commented on source/dynamical_systems/include/dynamical_systems/IDynamicalSystem.hpp:
```diff
@@ -0,0 +1,114 @@
+   * @return The parameter, if it exists
+   */
+  template<typename T>
+  T get_parameter(const std::string& name);
```
I will do this in a future PR too, if needed

domire8 commented on source/dynamical_systems/include/dynamical_systems/IDynamicalSystem.hpp:
```diff
@@ -0,0 +1,114 @@
+   * @brief Set a parameter of the dynamical system.
+   * @param parameter The new parameter
+   */
+  virtual void set_parameter(const std::shared_ptr<state_representation::ParameterInterface>& parameter);
```
I agree that this will result in a better design, also postponing `validate_parameter` to the next PR

domire8 commented:
I think I addressed all the comments above, the most important commit is probably the last one, where i changed the set/get parameter stuff. It becomes increasingly difficult for me to imagine the consequences of those changes in the actual implementation of the DSs, so if there is nothing completely wrong it would be nice to merge soon and then I can address all further problems when I start to refactor the DSs.

eeberhard reviewed (approved):
Thanks for taking the time with all the comments. I think that addresses all of them on my side, and it's starting to look very good for the next steps!
